BUG: creating Categorical from pandas Index/Series with "object" dtype infers string

niruta25 opened the pull request:
- [x] closes #61778
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Preserve [object](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) dtype for categories when constructing [Categorical](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) from pandas objects

This PR fixes an inconsistency in how pandas infers the dtype of categories when constructing a [Categorical](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) from different input types:

When constructing a [Categorical](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) from a pandas [Series](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) or [Index](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) with [dtype="object"](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1), the categories' dtype is now preserved as [object](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1).
When constructing from a NumPy array with [dtype="object"](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) or a raw Python sequence, pandas continues to infer the most specific dtype for the categories (e.g., [str](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) if all elements are strings).
This change brings the behavior of [Categorical](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) in line with how [Series](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) and [Index](https://github.com/pandas-dev/pandas/compare/main...niruta25:pandas:niruta_issue61778?expand=1) handle dtype preservation, making the API more consistent and predictable.

Example

```
pd.options.future.infer_string = True

ser = pd.Series(["foo", "bar", "baz"], dtype="object")
idx = pd.Index(["foo", "bar", "baz"], dtype="object")
arr = np.array(["foo", "bar", "baz"], dtype="object")
pylist = ["foo", "bar", "baz"]

cat_from_ser = pd.Categorical(ser)
cat_from_idx = pd.Categorical(idx)
cat_from_arr = pd.Categorical(arr)
cat_from_list = pd.Categorical(pylist)

# Series/Index with object dtype: preserve object dtype
assert cat_from_ser.categories.dtype == "object"
assert cat_from_idx.categories.dtype == "object"

# Numpy array or list: infer string dtype
assert cat_from_arr.categories.dtype == "str"
assert cat_from_list.categories.dtype == "str"
```

Documentation and release notes have been updated.
Closes: https://github.com/pandas-dev/pandas/issues/61778

niruta25 committed:
slack link update

niruta25 committed:
Merge pull request #2 from niruta25/niruta-#61690-slack

slack link update issue #61690

niruta25 committed:
Merge branch 'pandas-dev:main' into main

niruta25 committed:
Merge branch 'pandas-dev:main' into main

niruta25 committed:
Merge branch 'pandas-dev:main' into main

niruta25 committed:
Merge branch 'pandas-dev:main' into main

niruta25 committed:
Merge branch 'pandas-dev:main' into main

niruta25 committed:
object

niruta25 committed:
whatsnew

niruta25 commented on pandas/core/arrays/categorical.py:
```diff
@@ -465,15 +470,17 @@ def __init__(
                     codes, categories = factorize(values, sort=False)
                     if dtype.ordered:
-                        # raise, as we don't have a sortable data structure and so
-                        # the user should give us one by specifying categories
```
Although I do not have any strong preference, I am happy to add it back.

niruta25 committed:
some comments

niruta25 committed:
comment restore

niruta25 commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -180,6 +180,31 @@ def test_array_repr(self, data, size):
+
+        import pandas as pd
+
+        pd.options.future.infer_string = True
```
done

niruta25 committed:
assertionerror fix

niruta25 committed:
rst changes

niruta25 committed:
rst import error

niruta25 commented:
> > @jbrockmendel Regarding this [bug](https://github.com/pandas-dev/pandas/issues/61778), the change to always preserve object dtype for categories when constructing a Categorical from a pandas Series or Index with dtype="object" is a behavioral change that affects a wide range of pandas internals and user-facing APIs. Hence I am seeing a lot of failures.
> > I see two ways to resolve without changing overall behavior.
> > 
> > 1. Only Preserve object Dtype When All Elements Are Not Strings
> > 
> > 
> > * If the input is a pandas Series/Index with dtype="object", only preserve object dtype for categories if not all elements are strings.
> > * If all elements are strings, allow inference to str (the current behavior).
> > 
> > 
> > 2. Add a Keyword Argument to Categorical (e.g., preserve_object_dtype=False)
> > 
> > 
> > * Add an explicit option to the Categorical constructor to preserve the object dtype for categories.
> > * Default to the current behavior, but allow users to opt in to preservation.
> > 
> > Let me know your thoughts.
> 
> @jbrockmendel Thank you for your comments. I have addressed them. Although I doubt it would help address all the failing tests. Please let me know your thoughts. Any preference from above two options.

I tried out the 1st way and all the test cases are passing. Let me know your thoughts.

niruta25 committed:
Merge branch 'main' into niruta_issue61778

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -454,6 +454,11 @@ def __init__(
+                # Check for pandas Series/ Index with object dtye
+                preserve_object_dtpe = False
+                if isinstance(values, (ABCSeries, ABCIndex)):
+                    if getattr(values.dtype, "name", None) == "object":
```
you can just check values.dtype == object

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -470,7 +475,13 @@ def __init__(
                         ) from err
 
-                # we're inferring from values
+                # If we should prserve object dtype, force categories to object dtype
```
typo prserve -> preserve

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -470,7 +475,13 @@ def __init__(
+                # If we should prserve object dtype, force categories to object dtype
+                if preserve_object_dtpe:
+                    # Only preserve object dtype if not all elements are strings
+                    if not all(isinstance(x, str) for x in categories):
```
why is this check necessary?

niruta25 replied:
the change to always preserve object dtype for categories when constructing a Categorical from a pandas Series or Index with dtype="object" is a behavioral change that affects a wide range of pandas internals and user-facing APIs.
To make sure other functionality doesn't break, I am preserving object datatype when all elements are not string.

jbrockmendel replied:
the more i look at it, the more i think this misses the point of the motivating issue.  i mean, wouldn't the new test this PR adds pass on main?

niruta25 committed:
change condition

niruta25 commented:
> > @niruta25 - are you interested in continuing here? If not, I can pick this up.
> 
> Hey I would like to continue working in this if it's ok. Should resolve comments by this weekend.

@rhshadrach @jbrockmendel addressed the comments. Please review.

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -454,6 +454,11 @@ def __init__(
                 dtype = CategoricalDtype(categories, values.dtype.pyarrow_dtype.ordered)
             else:
+                # Check for pandas Series/ Index with object dtye
+                preserve_object_dtpe = False
```
typo dtpe -> dtype

jbrockmendel commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -861,7 +861,7 @@ Categorical
 - Bug in :meth:`DataFrame.pivot` and :meth:`DataFrame.set_index` raising an ``ArrowNotImplementedError`` for columns with pyarrow dictionary dtype (:issue:`53051`)
 - Bug in :meth:`Series.convert_dtypes` with ``dtype_backend="pyarrow"`` where empty :class:`CategoricalDtype` :class:`Series` raised an error or got converted to ``null[pyarrow]`` (:issue:`59934`)
--
+- Bug in :class:`Categorical` where constructing from a pandas :class:`Series` or :class:`Index` with ``dtype='object'`` did not preserve the categories' dtype as ``object``; now the dtype is preserved as ``object`` for these cases, while numpy arrays and Python sequences with ``dtype='object'`` continue to infer the most specific dtype (for example, ``str`` if all elements are strings).
```
issue ref at the end

jbrockmendel commented:
@niruta25 i updated your branch with the necessary changes in #62757.  If you'd like to pull those changes into this branch we can merge this instead.
